File: packages/fluent/src/runtime/run-store-adapter.ts
// @ts-nocheck -- Vendored TanStack source targets a looser optional-property TypeScript policy.
import type { DeleteReason, RunState, WorkflowEvent } from "@firegrid/core"
import type { WorkflowRunStoreAdapter, WorkflowRunStoreAdapterStore } from "./types"

export function createRunStoreAdapter(
  store: WorkflowRunStoreAdapterStore
): WorkflowRunStoreAdapter {
  return {
    getRunState(runId) {
      return store.loadRunState(runId)
    },

    setRunState(_runId: string, state: RunState) {
      return store.saveRunState({ state })
    },

    deleteRun(runId: string, reason: DeleteReason) {
      return store.deleteRun(runId, reason)
    },

    async appendEvent(
      runId: string,
      expectedNextIndex: number,
      event: WorkflowEvent
    ) {
      await store.appendEvents({
        runId,
        expectedNextIndex,
        events: [event]
      })
    },

    async getEvents(runId: string) {
      const events = await store.readEvents({ runId })
      return events.map((event) => event.event)
    },

    subscribe: store.subscribeEvents
      ? (runId, fromIndex, onEvent) => store.subscribeEvents!(runId, fromIndex, onEvent)
      : undefined
  }
}

File: src/Firegrid.Runtime/RunStoreAdapter.fs
namespace Firegrid.Runtime

open Fable.Core
open Firegrid.Core

/// `createRunStoreAdapter` — wraps a `WorkflowRunStoreAdapterStore` into Core's
/// `RunStore`.
[<RequireQualifiedAccess>]
module RunStoreAdapter =

    let create (store: WorkflowRunStoreAdapterStore) : WorkflowRunStoreAdapter =
        { GetRunState = fun runId -> store.LoadRunState runId

          SetRunState = fun _runId state -> store.SaveRunState { State = state }

          DeleteRun = fun runId reason -> store.DeleteRun runId reason

          AppendEvent =
            fun runId expectedNextIndex event ->
                promise {
                    let! _ =
                        store.AppendEvents
                            { RunId = runId
                              ExpectedNextIndex = expectedNextIndex
                              Events = [| event |] }

                    return ()
                }

          GetEvents =
            fun runId ->
                promise {
                    let! events = store.ReadEvents { RunId = runId; FromIndex = None }
                    return events |> Array.map (fun e -> e.Event)
                }

          Subscribe =
            match store.SubscribeEvents with
            | Some subscribeEvents ->
                Some(fun runId fromIndex onEvent -> subscribeEvents runId fromIndex onEvent)
            | None -> None }
